WalletWasabi/Blockchain/Transactions/SmartTransaction.cs

```diff
@@ -32,23 +35,52 @@ public SmartTransaction(Transaction transaction, Height height, uint256? blockHa
 		FirstSeen = firstSeen == default ? DateTimeOffset.UtcNow : firstSeen;
 
 		IsReplacement = isReplacement;
-		WalletInputs = new HashSet<SmartCoin>(Transaction.Inputs.Count);
-		WalletOutputs = new HashSet<SmartCoin>(Transaction.Outputs.Count);
+		_walletInputs = new HashSet<SmartCoin>(Transaction.Inputs.Count);
+		_walletOutputs = new HashSet<SmartCoin>(Transaction.Outputs.Count);
 	}
 
 	#endregion Constructors
 
 	#region Members
 
-	/// <summary>
-	/// Coins those are on the input side of the tx and belong to ANY loaded wallet. Later if more wallets are loaded this list can increase.
-	/// </summary>
-	public HashSet<SmartCoin> WalletInputs { get; }
+	public IReadOnlyCollection<SmartCoin> WalletInputs
```

Comment:
IDK if we rely on `HashSet`s ensuring of uniqueness. Did you check?

Reply:
I'm not sure if I understand exactly what you mean.

The uniqueness of `WalletInputs`'s members is still preserved, because
  * `WalletInputs = (IReadOnlyCollection) _walletInputs`, where `_walletInputs` has unique members because its type is `HashSet` and
  * `WalletInputs` is read only.